# Your task is to:
1. Perform Test Classification using Multinomial Naive Bayes(already implemented in sklearn).
2. Implement Naive Bayes on your own from scratch for text classification. 
3. Compare Results of your implementation of Naive Bayes with one in Sklearn.

Dataset - http://archive.ics.uci.edu/ml/datasets/Twenty+Newsgroups

In [1]:
import os
import pandas as pd
import re

#VOCABLURY.....

###path of folder
path_='D:/New folder/20_newsgroups'
list_=[]
i=0

###store all 20000 file's path in list
for subdir,dirs,files in os.walk(path_):
    for file in files:
        if i!=0:
            f=(os.path.join(subdir,file))  
            list_.append(f)
        i+=1    

###stop words- which are common in all files and give no meaning to our classification
stop_words=["a", "about", "above", "after", "again", "against", "ain", "all",'writes', "am", "an", "and", "any", "are", "aren", "aren't", "as", "at", "be", "because", "been", "before", "being", "below", "between", "both", "but", "by", "can", "couldn", "couldn't", "d", "did", "didn", "didn't", "do", "does", "doesn", "doesn't", "doing", "don", "don't", "down", "during", "each", "few", "for", "from", "further", "had", "hadn", "hadn't", "has", "hasn", "hasn't", "have", "haven", "haven't", "having", "he", "her", "here", "hers", "herself", "him", "himself", "his", "how", "i", "if", "in", "into", "is", "isn", "isn't", "it", "it's", "its", "itself", "just", "ll", "m", "ma", "me", "mightn", "mightn't", "more", "most", "mustn", "mustn't", "my", "myself", "needn", "needn't", "no", "nor", "not", "now", "o", "of", "off", "on", "once", "only", "or", "other", "our", "ours", "ourselves", "out", "over", "own", "re", "s", "same", "shan", "shan't", "she", "she's", "should", "should've", "shouldn", "shouldn't", "so", "some", "such", "t", "than", "that", "that'll", "the", "their", "theirs", "them", "themselves", "then", "there", "these", "they", "this", "those", "through", "to", "too", "under", "until", "up", "ve", "very", "was", "wasn", "wasn't", "we", "were", "weren", "weren't", "what", "when", "where", "which", "while", "who", "whom", "why", "will", "with", "won", "won't", "wouldn", "wouldn't", "y", "you", "you'd", "you'll", "you're", "you've", "your", "yours", "yourself", "yourselves", "could", "he'd", "he'll", "he's", "here's", "how's", "i'd", "i'll", "i'm", "i've", "let's", "ought", "she'd", "she'll", "that's", "there's", "they'd", "they'll", "they're", "they've", "we'd", "we'll", "we're", "we've", "what's", "when's", "where's", "who's", "why's", "would", "able", "abst", "accordance", "according", "accordingly", "across", "act", "actually", "added", "adj", "affected", "affecting", "affects", "afterwards", "ah", "almost", "alone", "along", "already", "also", "although", "always", "among", "amongst", "announce", "another", "anybody", "anyhow", "anymore", "anyone", "anything", "anyway", "anyways", "anywhere", "apparently", "approximately", "arent", "arise", "around", "aside", "ask", "asking", "auth", "available", "away", "awfully", "b", "back", "became", "become", "becomes", "becoming", "beforehand", "begin", "beginning", "beginnings", "begins", "behind", "believe", "beside", "besides", "beyond", "biol", "brief", "briefly", "c", "ca", "came", "cannot", "can't", "cause", "causes", "certain", "certainly", "co", "com", "come", "comes", "contain", "containing", "contains", "couldnt", "date", "different", "done", "downwards", "due", "e", "ed", "edu", "effect", "eg", "eight", "eighty", "either", "else", "elsewhere", "end", "ending", "enough", "especially", "et", "etc", "even", "ever", "every", "everybody", "everyone", "everything", "everywhere", "ex", "except", "f", "far", "ff", "fifth", "first", "five", "fix", "followed", "following", "follows", "former", "formerly", "forth", "found", "four", "furthermore", "g", "gave", "get", "gets", "getting", "give", "given", "gives", "giving", "go", "goes", "gone", "got", "gotten", "h", "happens", "hardly", "hed", "hence", "hereafter", "hereby", "herein", "heres", "hereupon", "hes", "hi", "hid", "hither", "home", "howbeit", "however", "hundred", "id", "ie", "im", "immediate", "immediately", "importance", "important", "inc", "indeed", "index", "information", "instead", "invention", "inward", "itd", "it'll", "j", "k", "keep", "keeps", "kept", "kg", "km", "know", "known", "knows", "l", "largely", "last", "lately", "later", "latter", "latterly", "least", "less", "lest", "let", "lets", "like", "liked", "likely", "line", "little", "'ll", "look", "looking", "looks", "ltd", "made", "mainly", "make", "makes", "many", "may", "maybe", "mean", "means", "meantime", "meanwhile", "merely", "mg", "might", "million", "miss", "ml", "moreover", "mostly", "mr", "mrs", "much", "mug", "must", "n", "na", "name", "namely", "nay", "nd", "near", "nearly", "necessarily", "necessary", "need", "needs", "neither", "never", "nevertheless", "new", "next", "nine", "ninety", "nobody", "non", "none", "nonetheless", "noone", "normally", "nos", "noted", "nothing", "nowhere", "obtain", "obtained", "obviously", "often", "oh", "ok", "okay", "old", "omitted", "one", "ones", "onto", "ord", "others", "otherwise", "outside", "overall", "owing", "p", "page", "pages", "part", "particular", "particularly", "past", "per", "perhaps", "placed", "please", "plus", "poorly", "possible", "possibly", "potentially", "pp", "predominantly", "present", "previously", "primarily", "probably", "promptly", "proud", "provides", "put", "q", "que", "quickly", "quite", "qv", "r", "ran", "rather", "rd", "readily", "really", "recent", "recently", "ref", "refs", "regarding", "regardless", "regards", "related", "relatively", "research", "respectively", "resulted", "resulting", "results", "right", "run", "said", "saw", "say", "saying", "says", "sec", "section", "see", "seeing", "seem", "seemed", "seeming", "seems", "seen", "self", "selves", "sent", "seven", "several", "shall", "shed", "shes", "show", "showed", "shown", "showns", "shows", "significant", "significantly", "similar", "similarly", "since", "six", "slightly", "somebody", "somehow", "someone", "somethan", "something", "sometime", "sometimes", "somewhat", "somewhere", "soon", "sorry", "specifically", "specified", "specify", "specifying", "still", "stop", "strongly", "sub", "substantially", "successfully", "sufficiently", "suggest", "sup", "sure", "take", "taken", "taking", "tell", "tends", "th", "thank", "thanks", "thanx", "thats", "that've", "thence", "thereafter", "thereby", "thered", "therefore", "therein", "there'll", "thereof", "therere", "theres", "thereto", "thereupon", "there've", "theyd", "theyre", "think", "thou", "though", "thoughh", "thousand", "throug", "throughout", "thru", "thus", "til", "tip", "together", "took", "toward", "towards", "tried", "tries", "truly", "try", "trying", "ts", "twice", "two", "u", "un", "unfortunately", "unless", "unlike", "unlikely", "unto", "upon", "ups", "us", "use", "used", "useful", "usefully", "usefulness", "uses", "using", "usually", "v", "value", "various", "'ve", "via", "viz", "vol", "vols", "vs", "w", "want", "wants", "wasnt", "way", "wed", "welcome", "went", "werent", "whatever", "what'll", "whats", "whence", "whenever", "whereafter", "whereas", "whereby", "wherein", "wheres", "whereupon", "wherever", "whether", "whim", "whither", "whod", "whoever", "whole", "who'll", "whomever", "whos", "whose", "widely", "willing", "wish", "within", "without", "wont", "words", "world", "wouldnt", "www", "x", "yes", "yet", "youd", "youre", "z", "zero", "a's", "ain't", "allow", "allows", "apart", "appear", "appreciate", "appropriate", "associated", "best", "better", "c'mon", "c's", "cant", "changes", "clearly", "concerning", "consequently", "consider", "considering", "corresponding", "course", "currently", "definitely", "described", "despite", "entirely", "exactly", "example", "going", "greetings", "hello", "help", "hopefully", "ignored", "inasmuch", "indicate", "indicated", "indicates", "inner", "insofar", "it'd", "keep", "keeps", "novel", "presumably", "reasonably", "second", "secondly", "sensible", "serious", "seriously", "sure", "t's", "third", "thorough", "thoroughly", "three", "well", "wonder", "a", "about", "above", "above", "across", "after", "afterwards", "again", "against", "all", "almost", "alone", "along", "already", "also", "although", "always", "am", "among", "amongst", "amoungst", "amount", "an", "and", "another", "any", "anyhow", "anyone", "anything", "anyway", "anywhere", "are", "around", "as", "at", "back", "be", "became", "because", "become", "becomes", "becoming", "been", "before", "beforehand", "behind", "being", "below", "beside", "besides", "between", "beyond", "bill", "both", "bottom", "but", "by", "call", "can", "cannot", "cant", "co", "con", "could", "couldnt", "cry", "de", "describe", "detail", "do", "done", "down", "due", "during", "each", "eg", "eight", "either", "eleven", "else", "elsewhere", "empty", "enough", "etc", "even", "ever", "every", "everyone", "everything", "everywhere", "except", "few", "fifteen", "fify", "fill", "find", "fire", "first", "five", "for", "former", "formerly", "forty", "found", "four", "from", "front", "full", "further", "get", "give", "go", "had", "has", "hasnt", "have", "he", "hence", "her", "here", "hereafter", "hereby", "herein", "hereupon", "hers", "herself", "him", "himself", "his", "how", "however", "hundred", "ie", "if", "in", "inc", "indeed", "interest", "into", "is", "it", "its", "itself", "keep", "last", "latter", "latterly", "least", "less", "ltd", "made", "many", "may", "me", "meanwhile", "might", "mill", "mine", "more", "moreover", "most", "mostly", "move", "much", "must", "my", "myself", "name", "namely", "neither", "never", "nevertheless", "next", "nine", "no", "nobody", "none", "noone", "nor", "not", "nothing", "now", "nowhere", "of", "off", "often", "on", "once", "one", "only", "onto", "or", "other", "others", "otherwise", "our", "ours", "ourselves", "out", "over", "own", "part", "per", "perhaps", "please", "put", "rather", "re", "same", "see", "seem", "seemed", "seeming", "seems", "serious", "several", "she", "should", "show", "side", "since", "sincere", "six", "sixty", "so", "some", "somehow", "someone", "something", "sometime", "sometimes", "somewhere", "still", "such", "system", "take", "ten", "than", "that", "the", "their", "them", "themselves", "then", "thence", "there", "thereafter", "thereby", "therefore", "therein", "thereupon", "these", "they", "thickv", "thin", "third", "this", "those", "though", "three", "through", "throughout", "thru", "thus", "to", "together", "too", "top", "toward", "towards", "twelve", "twenty", "two", "un", "under", "until", "up", "upon", "us", "very", "via", "was", "we", "well", "were", "what", "whatever", "when", "whence", "whenever", "where", "whereafter", "whereas", "whereby", 'takes',"wherein", "whereupon", "wherever", "whether", "which", "while", "whither", "who", "whoever", "whole", "whom", "whose", "why", "will", "with", "within", "without", "would", "yet", "you", "your", "yours", "yourself", "yourselves", "the", "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "co", "op", "research-articl", "pagecount", "cit", "ibid", "les", "le", "au", "que", "est", "pas", "vol", "el", "los", "pp", "u201d", "well-b", "http", "volumtype", "par", "0o", "0s", "3a", "3b", "3d", "6b", "6o", "a1", "a2", "a3", "a4", "ab", "ac", "ad", "ae", "af", "ag", "aj", "al", "an", "ao", "ap", "ar", "av", "aw", "ax", "ay", "az", "b1", "b2", "b3", "ba", "bc", "bd", "be", "bi", "bj", "bk", "bl", "bn", "bp", "br", "bs", "bt", "bu", "bx", "c1", "c2", "c3", "cc", "cd", "ce", "cf", "cg", "ch", "ci", "cj", "cl", "cm", "cn", "cp", "cq", "cr", "cs", "ct", "cu", "cv", "cx", "cy", "cz", "d2", "da", "dc", "dd", "de", "df", "di", "dj", "dk", "dl", "do", "dp", "dr", "ds", "dt", "du", "dx", "dy", "e2", "e3", "ea", "ec", "ed", "ee", "ef", "ei", "ej", "el", "em", "en", "eo", "ep", "eq", "er", "es", "et", "eu", "ev", "ex", "ey", "f2", "fa", "fc", "ff", "fi", "fj", "fl", "fn", "fo", "fr", "fs", "ft", "fu", "fy", "ga", "ge", "gi", "gj", "gl", "go", "gr", "gs", "gy", "h2", "h3", "hh", "hi", "hj", "ho", "hr", "hs", "hu", "hy", "i", "i2", "i3", "i4", "i6", "i7", "i8", "ia", "ib", "ic", "ie", "ig", "ih", "ii", "ij", "il", "in", "io", "ip", "iq", "ir", "iv", "ix", "iy", "iz", "jj", "jr", "js", "jt", "ju", "ke", "kg", "kj", "km", "ko", "l2", "la", "lb", "lc", "lf", "lj", "ln", "lo", "lr", "ls", "lt", "m2", "ml", "mn", "mo", "ms", "mt", "mu", "n2", "nc", "nd", "ne", "ng", "ni", "nj", "nl", "nn", "nr", "ns", "nt", "ny", "oa", "ob", "oc", "od", "of", "og", "oi", "oj", "ol", "om", "on", "oo", "oq", "or", "os", "ot", "ou", "ow", "ox", "oz", "p1", "p2", "p3", "pc", "pd", "pe", "pf", "ph", "pi", "pj", "pk", "pl", "pm", "pn", "po", "pq", "pr", "ps", "pt", "pu", "py", "qj", "qu", "r2", "ra", "rc", "rd", "rf", "rh", "ri", "rj", "rl", "rm", "rn", "ro", "rq", "rr", "rs", "rt", "ru", "rv", "ry", "s2", "sa", "sc", "sd", "se", "sf", "si", "sj", "sl", "sm", "sn", "sp", "sq", "sr", "ss", "st", "sy", "sz", "t1", "t2", "t3", "tb", "tc", "td", "te", "tf", "th", "ti", "tj", "tl", "tm", "tn", "tp", "tq", "tr", "ts", "tt", "tv", "tx", "ue", "ui", "uj", "uk", "um", "un", "uo", "ur", "ut", "va", "wa", "vd", "wi", "vj", "vo", "wo", "vq", "vt", "vu", "x1", "x2", "x3", "xf", "xi", "xj", "xk", "xl", "xn", "xo", "xs", "xt", "xv", "xx", "y2", "yj", "yl", "yr", "ys", "yt", "zi", "zz",'xref','date','lines','path','newsgroups','from','subject','message-id','sender','reply-to','reads','organization','references']

###store frequency of all words in dictionary
###all possible words in 20000 files-vocablary
d3={}
for i in list_:
    file_name=open(i,encoding='iso-8859-1')
    file_read=file_name.read().lower()
    st=re.split(',|;|\n|:| ',file_read)
    for j in st:
        if j.isalpha() and j not in stop_words:
            d3[j]=d3.get(j,0)+1    

###pick top k words for vocablury
d_main=pd.DataFrame(d3.items())
d_main.sort_values(1,ascending=False,inplace=True)
d_main=d_main.reset_index(drop=True)
d_main2=d_main[:2500]
features=d_main2[0]

In [2]:
###create a DATAFRAME
###columns-vocablury
###rows-20000 files
matrix=pd.DataFrame(index=list_,columns=features).fillna(0)
matrix=matrix.reset_index(drop=True)

In [3]:
###fill values in dataframe--no. of times a word is present in particular document
for i in range(len(list_)):
    file_open=open(list_[i],encoding='iso-8859-1')
    file_r=file_open.read().lower()
    ###split string by multiple delimiters
    string=re.split(',|;| ',file_r)
    for j in string:
        for k in features:
            if j==k:
                matrix[k][i]+=1
matrix                

0      apr  gmt  article  university  people  good  time  distribution  god  \
0        2    0        1           0       2     0     0             0    0   
1        1    0        0           0       1     0     0             0    0   
2        1    0        0           0       0     0     0             0    0   
3        1    0        2           0       0     0     0             0    0   
4        0    0        0           0       2     0     0             0    0   
...    ...  ...      ...         ...     ...   ...   ...           ...  ...   
19992    1    0        1           0       1     0     0             0    0   
19993    1    0        1           0       0     0     0             0    0   
19994    1    0        1           0       1     0     0             0    0   
19995    1    0        2           0       0     0     0             0    0   
19996    1    0        3           0       1     2     0             0    0   

0      computer  ...  county  accident  pieces  hidden  fell  null  square  \
0             1  ...       0         0       0       0     0     0       0   
1             0  ...       0         0       0       0     0     0       0   
2             0  ...       0         0       0       0     0     0       0   
3             0  ...       0         0       0       0     0     0       0   
4             0  ...       0         0       0       0     0     0       0   
...         ...  ...     ...       ...     ...     ...   ...   ...     ...   
19992         0  ...       0         0       0       0     0     0       0   
19993         0  ...       0         0       0       0     0     0       0   
19994         0  ...       0         0       0       0     0     0       0   
19995         0  ...       0         0       0       0     0     0       0   
19996         0  ...       0         1       0       0     0     0       0   

0      raid  unusual  extent  
0         0        0       0  
1         0        0       0  
2         0        0       0  
3         0        0       0  
4         0        0       0  
...     ...      ...     ...  
19992     0        0       0  
19993     0        0       0  
19994     0        0       0  
19995     0        0       0  
19996     0        0       0  

[19997 rows x 2500 columns]

In [4]:
###define class for each document
import numpy as np
list4=[]
for i in list_:
    list4.append(i.split('/')[-2])
dt={}
t=0
for i in list4:
    if i not in dt:
        dt[i]=t
        t+=1
output=np.zeros(len(list4))
for i in range(len(list4)):
    output[i]=dt[list4[i]]
output

array([ 0.,  0.,  0., ..., 19., 19., 19.])

In [5]:
#SCORE
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
import numpy as np

x_train,x_test,y_train,y_test=train_test_split(matrix,output,random_state=0,test_size=0.2)
clf=MultinomialNB()
clf.fit(x_train,y_train)
print(clf.score(x_train,y_train))
print(clf.score(x_test,y_test))

0.7998999812464838
0.7435


# IMPLEMENTATION

In [6]:
x_train=x_train.reset_index(drop=True)
x_test=x_test.reset_index(drop=True)

In [7]:
#features
features=x_train.columns

In [8]:
#convert training and testing into numpy array for better calculation
x_train=np.array(x_train)
x_test=np.array(x_test)

In [9]:
x_train=x_train[:,1:]
x_test=x_test[:,1:]

In [25]:
#fit function 
def fit(x_train,y_train,features):
    dictionary={}
    dictionary['total_data']=len(y_train)
    x1=list(set(y_train))
    for i in x1:
        dictionary[i]={}
        new=np.where(y_train==i)
        x_train_current=x_train[new]
        for j in range(len(features)-1):
            dictionary[i][features[j]]=x_train_current[:,j].sum()
        p=list(dictionary[i].values())
        dictionary[i]['total_words']=sum(p)-len(x_train_current)  
        dictionary[i]['class_p']=len(x_train_current)
    return dictionary               

In [26]:
fit_dict=fit(x_train,y_train,features)

In [37]:
#CALCULATE PROBABILITY FOR EVERY CLASS AND PREDICTSINGLEPOINT COMPARES THE ACCURACY FOR EVERY DOCUMENT AND PICK THE CLASS WITH MAXIMUM PROBABILITY
def probability(dictionary,x,current_class,features):
    output=np.log(dictionary[current_class]['class_p'])-np.log(dictionary['total_data'])
    for i in range(len(features)-1):
        classi=dictionary[current_class]['total_words']+len(features)-2
        if features[i]=='total_words' or features[i]=='class_p' or x[i]==0:
            continue   
        pwi=dictionary[current_class][features[i]]+1
        output+=(np.log(pwi)-np.log(classi))
    return output    

In [38]:
###CALCULATE BEST CLASS FOR EVERY SINGLE DOCUMENT IN TEST DATA

def predictsingleword(fit_dict,x,features):
    classes=fit_dict.keys()
    best_p=-10000
    best_class=-1
    first_run=True
    for i in classes:
        if i=='total_data':
            continue
        p_current_class=probability(fit_dict,x,i,features)
        if (first_run or p_current_class>best_p):
            best_p=p_current_class
            best_class=i
            
        first_run=False
    return best_class    

In [39]:
#PREDICT CLASS FOR EVERY DOCUMENT WITH THE HELP OF predictsinglepoint and probability FUNCTIONS
import re
def predict(fit_dict,x_test,features):
    y_pred=[]
    for i in range(len(x_test)):
        pred_class=predictsingleword(fit_dict,x_test[i],features)
        y_pred.append(pred_class)
    return y_pred     

In [40]:
###PREDICTED ARRAY FOR TEST DATA
ans=predict(fit_dict,x_test,features)

In [42]:
###CALCULATE SCORE OF ACTUAL TEST DATA AND PREDICTED TEST DATA
from sklearn.metrics import accuracy_score
accuracy_score(y_test,ans)

0.737

In [45]:
from sklearn.metrics import classification_report,confusion_matrix
print(classification_report(y_test,ans))
print(confusion_matrix(y_test,ans))

              precision    recall  f1-score   support

         0.0       0.89      0.78      0.83       193
         1.0       0.74      0.76      0.75       201
         2.0       0.66      0.70      0.68       197
         3.0       0.58      0.73      0.65       182
         4.0       0.88      0.83      0.85       196
         5.0       0.65      0.75      0.70       196
         6.0       0.89      0.91      0.90       214
         7.0       0.93      0.92      0.93       221
         8.0       0.92      0.75      0.83       224
         9.0       0.56      0.62      0.59       194
        10.0       0.79      0.86      0.82       191
        11.0       0.75      0.67      0.71       194
        12.0       0.64      0.61      0.63       196
        13.0       0.74      0.60      0.66       210
        14.0       0.63      0.71      0.67       195
        15.0       0.70      0.81      0.75       200
        16.0       0.86      0.83      0.85       202
        17.0       0.81    

# COMPARISION....

score calculated by inbuilt multinomial naive bayes(MNB) function is 0.7435
score calculated by functions designed by us(self implementation) is 0.737
As we can see the accuracy is nearly closing to the multinomial naive bayes so we can infer that our code for self implementation is nearly closing to perfection
Hence implied...